pyelftool_parser/src/analyzer.py

```diff
@@ -0,0 +1,203 @@
+import sys
+import register
+import execute
+from debug import loginst, log, logresult, logstack
+from capstone.x86 import *
+from elftools.elf.elffile import ELFFile
+from capstone import *
+from elftools.elf.sections import (
+    NoteSection, SymbolTableSection, SymbolTableIndexSection
+)
+class disassembler:
+    def __init__(self, path):
+        self.path = path
+        self.inst = dict()
+        self.symbol = dict()
+        self.vertex = dict()
+        self.entry_pc = 0
+        self.exit_pc = 0
+        self.acquire_stack_address = 0
+
+    def disasminst(self):
+        pc_flag = 0
+        with open(self.path, 'rb') as f:
+            elf = ELFFile(f)
+            code = elf.get_section_by_name('.text')
+            ops = code.data()
+            addr = code['sh_addr']
+            md = Cs(CS_ARCH_X86, CS_MODE_64)
+            md.detail = True
+            for i in md.disasm(ops, addr):
+                self.inst[i.address] = (i)
+                if (i.address in self.symbol):  ## it is to catch the next symbol start. @minghwu: I think it could have a better way to do this.
+                    pc_flag = 0
+                if (i.address == self.entry_pc): ## when it is entry point, set the flag = 1 to catch exitpoint.
+                    pc_flag = 1
+                if (pc_flag == 1):               ## catch the point until the next symbol, means it is exit point.
+                    self.exit_pc = i.address
+                log(f'0x{i.address:x}:\t{i.mnemonic}\t{i.op_str}')
+
+    def disasmsymbol(self):
+        with open(self.path, 'rb') as f:
+            e = ELFFile(f)
+            symbol_tables = [ s for s in e.iter_sections()
+                         if isinstance(s, SymbolTableSection)]
+            for section in symbol_tables:
+                for symbol in section.iter_symbols():
+                    self.symbol[symbol['st_value']] = symbol.name
+                    self.vertex[symbol['st_value']] = symbol.name
+                    log("here")
+                    log(symbol.name, symbol['st_value'])
+                    if(symbol.name == '__cosrt_upcall_entry'): ## set up the entry pc.
+                        self.entry_pc = symbol['st_value']
+                        log("Set up entry point")
+                        log(hex(self.entry_pc))
+                    if(symbol.name == 'custom_acquire_stack'):
```

Comment:
Am I understanding correctly that you don't treat the `__cosrt_s_*` symbols as entry points? Is the logic that they each call `custom_acquire_stack`, so it is OK to start tracing from there? That might miss some stack operations. I think that `__cosrt_upcall_entry` also uses that symbol, so you might do redundant computations.

I'm likely just missing something here ;-)

Reply:
Sorry for that. I just test the __cosrt_upcall_entry. I will give it a look.